### PR TITLE
[IMP] website_payment: filter payment providers in portal

### PR DESCRIPTION
**Pull request by vchu-odoo**
Before from the portal the user could see all the payment providers even if the payment provider was not available from the website where the order was made. Now the payment providers are filtered if the order has website_id.

task-3284582

### CODE REVIEW

File: addons/website_payment/models/payment_provider.py [hunk 1]
```diff
@@ -25,6 +25,11 @@ def _get_compatible_providers(self, *args, website_id=None, **kwargs):
         :rtype: payment.provider
         """
         providers = super()._get_compatible_providers(*args, website_id=website_id, **kwargs)
+        # retrieve website_id from the order if it is defined for the correct portal view
+        if not website_id and kwargs.get('sale_order_id'):
+            order = request.env['sale.order'].sudo().browse(kwargs.get('sale_order_id')).exists()
```
Comment (Demesmaeker) [inline addons/website_payment/models/payment_provider.py:30]: When sudoing, it's better to indicated it in the variable name, here `order_sudo` or `order_su`. This way, it's easier to remember the order is in sudo, and we need to be careful to what we are doing with it.

**Review by Demesmaeker — CHANGES_REQUESTED** [review]

**Review by Demesmaeker — APPROVED** [review]

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Although the proposed diff is correct, I think it'd be better to inject `website_id` as early as possible in the chain of methods receiving and passing over kwargs, rather than having `_get_compatible_provider` figure out the `sale_order_id` (just in case there is one) in the `website_payment` that has no dependency on `sale`.

See below:

```patch
diff --git a/addons/website_sale/controllers/main.py b/addons/website_sale/controllers/main.py
index 090c1a41f39..48f8abe9f01 100644
--- a/addons/website_sale/controllers/main.py
+++ b/addons/website_sale/controllers/main.py
@@ -1841,6 +1841,19 @@ class PaymentPortal(payment_portal.PaymentPortal):
 
 
 class CustomerPortal(sale_portal.CustomerPortal):
+
+    def _get_payment_values(self, order_sudo, website_id=None, **kwargs):
+        """ Override of `sale` to inject the `website_id` into the kwargs.
+
+        :param sale.order order_sudo: The sales order being paid.
+        :param int website_id: The website on which the order was made, if any, as a `website` id.
+        :param dict kwargs: Locally unused keywords arguments.
+        :return: The payment-specific values.
+        :rtype: dict
+        """
+        website_id = website_id or order_sudo.website_id.id
+        return super()._get_payment_values(order_sudo, website_id=website_id, **kwargs)
+
     def _sale_reorder_get_line_context(self):
         return {}
```

**Comment by vchu-odoo** [comment]
> Although the proposed diff is correct, I think it'd be better to inject `website_id` as early as possible in the chain of methods receiving and passing over kwargs, rather than having `_get_compatible_provider` figure out the `sale_order_id` (just in case there is one) in the `website_payment` that has no dependency on `sale`.
> 
> See below:
> 
> ```diff
> diff --git a/addons/website_sale/controllers/main.py b/addons/website_sale/controllers/main.py
> index 090c1a41f39..48f8abe9f01 100644
> --- a/addons/website_sale/controllers/main.py
> +++ b/addons/website_sale/controllers/main.py
> @@ -1841,6 +1841,19 @@ class PaymentPortal(payment_portal.PaymentPortal):
>  
>  
>  class CustomerPortal(sale_portal.CustomerPortal):
> +
> +    def _get_payment_values(self, order_sudo, website_id=None, **kwargs):
> +        """ Override of `sale` to inject the `website_id` into the kwargs.
> +
> +        :param sale.order order_sudo: The sales order being paid.
> +        :param int website_id: The website on which the order was made, if any, as a `website` id.
> +        :param dict kwargs: Locally unused keywords arguments.
> +        :return: The payment-specific values.
> +        :rtype: dict
> +        """
> +        website_id = website_id or order_sudo.website_id.id
> +        return super()._get_payment_values(order_sudo, website_id=website_id, **kwargs)
> +
>      def _sale_reorder_get_line_context(self):
>          return {}
> ```

Yes, that is indeed a better solution, thank you!

**Review by AntoineVDV — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**